File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        String s;
        Stack<Character> stack;
        boolean isNo;

        for(int i=0; i<t; i++) {
            stack = new Stack<>();
            s = br.readLine();
            isNo = false;
            for(int j=0; j<s.length(); j++) {
                char c = s.charAt(j);
                if(c == '(') {
                    stack.push(c);
                } else {
                    if(stack.isEmpty()) {
                        System.out.println("NO");
                        isNo = true;
                        break;
                    }
                    stack.pop();
                }
            }

            if(!isNo) {
                if(stack.isEmpty()) System.out.println("YES");
                else System.out.println("NO");
            }
        }
    }
}

File: 백준/Silver/9012. 괄호/괄호.kt
import java.io.*

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()
    val sb = StringBuilder()
    repeat(n) {
        val line = br.readLine()
        val stack = ArrayDeque<Char>()
        var isVPS = true
        for (c in line) {
            when (c) {
                '(' -> stack.addLast(c)
                ')' -> {
                    if (stack.isNotEmpty() && stack.last() == '(') {
                        stack.removeLast()
                    } else {
                        isVPS = false
                        break
                    }
                }
            }
        }
        if (stack.isNotEmpty()) isVPS = false
        sb.appendLine(if (isVPS) "YES" else "NO")
    }
    print(sb)
}